### DISCUSSION TITLE
A variable length array record as one record

### DISCUSSION THREAD

**Discussion by Leg501** [DavidAce/h5pp#17] [Show and tell]
Good afternoon , I'm trying to figure out how to write and read a variable length set in 1 entry. What I mean. I have a pointer to a common data pool array of bytes uint8 and a list of segment sizes from this data pool. I want to save each segment as a single item.

std::vector<uint8_t> pool{0x35,0x34, 0x35, 0x00, 0x00, 0x35,0x36,0x36, 0x36, 0x36, 0x36, 0x00, 0x36, 0x00, 0x36, 0x36, 0x36 };
std::vector<int>sizeseg{4,5,3,5}

waiting in the file when saving:
0x35 0x34 0x35 0x00
0x00, 0x35,0x36,0x36, 0x36
0x36, 0x36, 0x00
0x36, 0x00, 0x36, 0x36, 0x36

When you read it, you can reconstruct it into a set of vectors if it is possible
std::vector<vector<uint8_t>>

**Comment by DavidAce** [comment]
Hello, 
You can write variable-length data using the `hvl_t` type provided by HDF5. It is essentially a struct with 2 members:

```c++
  /**
   * Variable Length Datatype struct in memory (This is only used for VL
   * sequences, not VL strings, which are stored in char *'s)
   */
  typedef struct {
      size_t len; /**< Length of VL data (in base type units) */
      void  *p;   /**< Pointer to VL data */
  } hvl_t;
```
You can then populate a buffer in the form of `std::vector<hvl_t>` where each member covers the segments in your original buffer `std::vector<uint8_t>`. 

To create the HDF5 datatype you need to use `H5Tvlen_create(H5T_NATIVE_UINT8)` and pass this along when you create the dataset. 
See for instance [example 04c](https://github.com/DavidAce/h5pp/blob/5c98a6f24c50f0e2c2b88ce4b4d954201a97f664/examples/example-04c-compound-datatype-vlen-arrays.cpp), how the `hvl_t year` member is treated.
